streaming_timeseries/streaming_timeseries.ipynb

```diff
@@ -0,0 +1,948 @@
+    "\n",
+    "        # Reset buffer when channel configuration changes\n",
+    "        if hasattr(self, \"buffer\"):\n",
+    "            self.buffer.clear()\n",
+    "            self.buffer.data = self.initial_data()\n",
+    "\n",
+    "    def initial_data(self) -> pd.DataFrame:\n",
+    "        # Create empty DataFrame with time and channel columns\n",
+    "        return pd.DataFrame({\"time\": [], **{ch: [] for ch in self.channel_names}})\n",
+    "\n",
+    "    def create_widgets(self) -> None:\n",
+    "        # Data source selection dropdown\n",
+    "        self.data_source_widget = pn.widgets.Select(\n",
+    "            name=\"Data Source\",\n",
+    "            options=self.data_source_names,\n",
+    "            value=self.data_source,\n",
+    "        )\n",
+    "        self.data_source_widget.param.watch(self.on_data_source_change, \"value\")\n",
+    "\n",
+    "        # Stream control radio buttons with custom styling for active state\n",
+    "        self.radio_group = pn.widgets.RadioButtonGroup(\n",
+    "            name=\"Stream Control\",\n",
+    "            options=[\"Start\", \"Pause\", \"Stop\"],\n",
+    "            value=\"Stop\",\n",
+    "            button_type=\"default\",\n",
+    "            sizing_mode=\"stretch_width\",\n",
+    "            stylesheets=[\n",
+    "                \"\"\"\n",
+    "                :host(.solid) .bk-btn.bk-btn-default.bk-active {\n",
+    "                    background-color: #b23c3c;\n",
+    "                }\n",
+    "            \"\"\"\n",
+    "            ],\n",
+    "        )\n",
+    "        self.radio_group.param.watch(self.handle_state_change, \"value\")\n",
+    "\n",
+    "    def on_data_source_change(self, event) -> None:\n",
+    "        # Stop any existing stream before switching data sources\n",
+    "        time.sleep(0.1)\n",
+    "        self.stop_stream()\n",
+    "        self.data_source = event.new\n",
+    "        self.update_channel_names()\n",
+    "\n",
+    "        # Reset visualization with empty plots\n",
+    "        self.main_streaming_pane.object = self.bare_stream_plot()\n",
+    "        self.position_pane.object = self.bare_pos_plot()\n",
+    "\n",
+    "    def handle_state_change(self, event) -> None:\n",
+    "        # Route radio button selections to appropriate stream control methods\n",
+    "        if event.new == \"Start\":\n",
+    "            self.start_stream()\n",
+    "        elif event.new == \"Pause\":\n",
+    "            self.pause_stream()\n",
+    "        elif event.new == \"Stop\":\n",
+    "            self.stop_stream()\n",
+    "\n",
+    "    def create_layout(self) -> None:\n",
+    "        # Initialize main visualization panes with empty plots\n",
+    "        self.position_pane = pn.pane.HoloViews(self.bare_pos_plot())\n",
+    "        self.main_streaming_pane = pn.pane.HoloViews(self.bare_stream_plot())\n",
+    "\n",
+    "        if self.notebook:\n",
```

Comment:
I think this is a great suggestion. I didn't even know it existed, but it's definitely in the documentation https://panel.holoviz.org/api/state.html. Do you have examples that use this feature? If so, I think it would be nice to show some links to it in this notebook.

Reply:
I use `pn.state.served` quite a lot in some code bases for the code to follow a different branch, usually to distinguish between the app being served and the test suite running.

Though I just remembered there were some subtleties with it (see more in https://github.com/holoviz/panel/issues/5623). Since no other example use it and it's yet another thing to explain in a rather complex example, I'd say forget about my suggestion and let's stick with the more explicit `notebook` attribute?

Reply:
Sure, although I would personally like more info into its use as it looks like it could be a great flag to use. I'll see if I can put something on our calendar.

Reply:
yea, using a notebook flag is super hacky and I'd much prefer if there was a way to detect with pn.state.served, however, given the timeline, let's just come back to this in a later PR